src/NuGet.Core/NuGet.ProjectModel/Utf8JsonStreamReader.cs

```diff
@@ -0,0 +1,273 @@
+
+namespace NuGet.ProjectModel
+{
+    /// <summary>
+    /// This struct is used to read over a memeory stream in parts, in order to avoid reading the entire stream into memory.
+    /// It functions as a wrapper around <see cref="Utf8JsonStreamReader"/>, while maintaining a stream and a buffer to read from.
+    /// </summary>
+    internal ref struct Utf8JsonStreamReader
+    {
+        private static readonly char[] DelimitedStringDelimiters = [' ', ','];
+        private static readonly byte[] Utf8Bom = [0xEF, 0xBB, 0xBF];
+
+        private const int BufferSizeDefault = 16 * 1024;
+        private const int MinBufferSize = 1024;
+        private Utf8JsonReader _reader;
+#pragma warning disable CA2213 // Disposable fields should be disposed
+        private Stream _stream;
+#pragma warning restore CA2213 // Disposable fields should be disposed
+        // The buffer is used to read from the stream in chunks.
+        private byte[] _buffer;
+        private bool _disposed;
+        private ArrayPool<byte> _bufferPool;
+        private int _bufferUsed = 0;
+
+        internal Utf8JsonStreamReader(Stream stream, int bufferSize = BufferSizeDefault, ArrayPool<byte> arrayPool = null)
+        {
+            if (stream is null)
+            {
+                throw new ArgumentNullException(nameof(stream));
+            }
+
+            if (bufferSize < MinBufferSize)
+            {
+                throw new ArgumentException($"Buffer size must be at least {MinBufferSize} bytes", nameof(bufferSize));
+            }
+
+            _bufferPool = arrayPool ?? ArrayPool<byte>.Shared;
+            _buffer = _bufferPool.Rent(bufferSize);
+            _disposed = false;
+            _stream = stream;
+            _stream.Read(_buffer, 0, 3);
+            if (!Utf8Bom.AsSpan().SequenceEqual(_buffer.AsSpan(0, 3)))
+            {
+                _bufferUsed = 3;
+            }
+
+            var iniialJsonReaderState = new JsonReaderState(new JsonReaderOptions
+            {
+                AllowTrailingCommas = true,
+                CommentHandling = JsonCommentHandling.Skip,
+            });
+
+            ReadStreamIntoBuffer(iniialJsonReaderState);
+            _reader.Read();
+        }
+
+        internal bool IsFinalBlock => _reader.IsFinalBlock;
+
+        internal JsonTokenType TokenType => _reader.TokenType;
+
+        internal bool ValueTextEquals(ReadOnlySpan<byte> utf8Text) => _reader.ValueTextEquals(utf8Text);
+
+        internal bool TryGetInt32(out int value) => _reader.TryGetInt32(out value);
+
+        internal string GetString() => _reader.GetString();
+
+        internal bool GetBoolean() => _reader.GetBoolean();
+
+        internal int GetInt32() => _reader.GetInt32();
+
+        internal bool Read()
+        {
+            ThrowExceptionIfDisposed();
+
+            bool wasRead;
+            while (!(wasRead = _reader.Read()) && !_reader.IsFinalBlock)
+            {
+                GetMoreBytesFromStream();
+            }
```

Comment:
In some cases, which I currently don't understand, the STJ implementation raises an exception when `IsFinalBlock` is set to true. Therefore, it would be better to check this value first before invoking the underlying `Read` method. Another advantage is that reading `IsFinalBlock` is an O(1) operation.

```suggestion
            while (!_reader.IsFinalBlock && !(wasRead = _reader.Read()))
            {
                GetMoreBytesFromStream();
            }
```



https://github.com/dotnet/runtime/blob/3a5bea5d60ea04b897ac968a358ca99a1189d368/src/libraries/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs#L269-L289

Reply:
If we check for the final block first then we won't all the read at all. The exception you're seeing there means that the Utf8JsonReader was told it has all the data but the current property is [none](https://learn.microsoft.com/en-us/dotnet/api/system.text.json.jsontokentype?view=net-8.0#fields). None meaning that there is no JSON data in the reader. If we want to check for the scenario we can, I would think it should be in the constructor though not here.